yellowbrick/utils/kneed.py

```diff
@@ -0,0 +1,208 @@
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+# IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# ID: kneed.py [] pswaldia@no-reply.github.com $
+
+"""
+This package contains a port of the knee-point detection package, kneed, by 
+Kevin Arvai and hosted at https://github.com/arvkevi/kneed. This port is maintained 
+with permission by the Yellowbrick contributors.
+"""
+import numpy as np
+from scipy import interpolate
+from scipy.signal import argrelextrema
+import warnings
+
+
+class KneeLocator(object):
+
+    def __init__(self, x, y, S=1.0, curve='concave', direction='increasing'):
+        """
+        Once instantiated, this class attempts to find the point of maximum
+        curvature on a line. The knee is accessible via the `.knee` attribute.
+        :param x: x values.
+        :type x: list or array.
+        :param y: y values.
+        :type y: list or array.
+        :param S: Sensitivity, original paper suggests default of 1.0
+        :type S: float
+        :param curve: If 'concave', algorithm will detect knees. If 'convex', it
+            will detect elbows.
+        :type curve: string
+        :param direction: one of {"increasing", "decreasing"}
+        :type direction: string
+        """
+        # Step 0: Raw Input
+        self.x = x
+        self.y = y
+        self.curve = curve
+        self.direction = direction
+        self.N = len(self.x)
+        self.S = S
+
+        # Step 1: fit a smooth line
+        uspline = interpolate.interp1d(self.x, self.y)
+        self.Ds_x = np.linspace(np.min(self.x), np.max(self.x), self.N)
+        self.Ds_y = uspline(self.Ds_x)
+
+        # Step 2: normalize values
+        self.xsn = self.__normalize(self.Ds_x)
+        self.ysn = self.__normalize(self.Ds_y)
+
+        # Step 3: Calculate difference curve
+        self.xd = self.xsn
+        if self.curve == 'convex' and direction == 'decreasing':
+            self.yd = self.ysn + self.xsn
+            self.yd = 1 - self.yd
+        elif self.curve == 'concave' and direction == 'decreasing':
+            self.yd = self.ysn + self.xsn
+        elif self.curve == 'concave' and direction == 'increasing':
+            self.yd = self.ysn - self.xsn
+        if self.curve == 'convex' and direction == 'increasing':
+            self.yd = abs(self.ysn - self.xsn)
+
+        # Step 4: Identify local maxima/minima
+        # local maxima
+        self.xmx_idx = argrelextrema(self.yd, np.greater)[0]
+        self.xmx = self.xd[self.xmx_idx]
+        self.ymx = self.yd[self.xmx_idx]
+
+        # local minima
+        self.xmn_idx = argrelextrema(self.yd, np.less)[0]
+        self.xmn = self.xd[self.xmn_idx]
+        self.ymn = self.yd[self.xmn_idx]
+
+        # Step 5: Calculate thresholds
+        self.Tmx = self.__threshold(self.ymx)
+
+        # Step 6: find knee
+        self.knee, self.norm_knee, self.knee_x = self.find_knee()
```

Comment:
Note that `find_knee()` is called in `__init__` as mentioned above, therefore there is no need to call it again in the visualizer. It may be interesting to refactor this into the `fit()`, `predict()` API to make it like an estimator; perhaps we could add that as an issue for a later revision.

Reply:
@bbengfort I am trying to become more and more familiar with yb codebase and the style of writing. It will be very exciting to refactor this into fit(), predict() api. I would start work on that once this feature works.

Reply:
Ok, great - but could we do that in a follow-up PR so that we have the original port in a commit?